File: Packages/SecuPersoDomain/Sources/SecuPersoDomain/Protocols.swift
import Foundation

public protocol ExposureMonitoringService: Sendable {
    func refresh() async throws -> [ExposureRecord]
    func stream() -> AsyncStream<[ExposureRecord]>
}

public protocol LoginActivityService: Sendable {
    func refresh() async throws -> [LoginEvent]
    func stream() -> AsyncStream<[LoginEvent]>
}

public protocol ProviderConnectionService: Sendable {
    func beginConnection(for provider: ProviderID) async -> AsyncStream<ProviderConnectionUpdate>
    func disconnect(_ provider: ProviderID) async throws
}

public protocol IncidentService: Sendable {
    func create(from loginEventID: UUID) async throws -> IncidentCase
    func resolve(_ incidentID: UUID) async throws
}

public protocol SecureStore: Sendable {
    func read(_ key: String) throws -> Data?
    func write(_ value: Data, for key: String) throws
    func delete(_ key: String) throws
}

public protocol IncidentReadableService: Sendable {
    func list() async throws -> [IncidentCase]
}

public protocol ProviderCatalogService: Sendable {
    func providers() async throws -> [ProviderDescriptor]
}

public protocol ProviderConnectionReadableService: Sendable {
    func connections() async throws -> [ProviderConnection]
}

public protocol ScenarioControlService: Sendable {
    func setScenario(_ scenario: FixtureScenario) async throws
    func currentScenario() async -> FixtureScenario
}

public protocol LoginEventActionService: Sendable {
    func markAsExpected(_ loginEventID: UUID) async throws -> LoginEvent?
}

public protocol ExposureSourceConfigurationService: Sendable {
    func loadConfiguration() async throws -> ExposureSourceConfiguration
    func saveConfiguration(_ configuration: ExposureSourceConfiguration) async throws
}

public protocol MonitoredEmailService: Sendable {
    func listMonitoredEmails() async throws -> [MonitoredEmailAddress]
    func addMonitoredEmail(_ email: String, providerHint: ProviderID) async throws -> MonitoredEmailAddress
    func setMonitoredEmailEnabled(id: UUID, isEnabled: Bool) async throws
    func removeMonitoredEmail(id: UUID) async throws
}

File: src/domain/protocols.ts
import type {
  ExposureRecord,
  LoginEvent,
  ProviderID,
  ProviderConnectionUpdate,
  ProviderConnection,
  IncidentCase,
  ProviderDescriptor,
  FixtureScenario,
  ExposureSourceConfiguration,
  MonitoredEmailAddress,
} from './models';

export interface ExposureMonitoringService {
  refresh(): Promise<ExposureRecord[]>;
  stream(callback: (values: ExposureRecord[]) => void): () => void;
}

export interface LoginActivityService {
  refresh(): Promise<LoginEvent[]>;
  stream(callback: (values: LoginEvent[]) => void): () => void;
}

export interface ProviderConnectionService {
  beginConnection(provider: ProviderID, onUpdate: (update: ProviderConnectionUpdate) => void): Promise<() => void>;
  disconnect(provider: ProviderID): Promise<void>;
}

export interface IncidentService {
  create(loginEventID: string): Promise<IncidentCase>;
  resolve(incidentID: string): Promise<void>;
}

export interface SecureStore {
  read(key: string): Promise<string | null>;
  write(key: string, value: string): Promise<void>;
  delete(key: string): Promise<void>;
}

export interface IncidentReadableService {
  list(): Promise<IncidentCase[]>;
}

export interface ProviderCatalogService {
  providers(): Promise<ProviderDescriptor[]>;
}

export interface ProviderConnectionReadableService {
  connections(): Promise<ProviderConnection[]>;
}

export interface ScenarioControlService {
  setScenario(scenario: FixtureScenario): Promise<void>;
  currentScenario(): Promise<FixtureScenario>;
}

export interface LoginEventActionService {
  markAsExpected(loginEventID: string): Promise<LoginEvent | null>;
}

export interface ExposureSourceConfigurationService {
  loadConfiguration(): Promise<ExposureSourceConfiguration>;
  saveConfiguration(config: ExposureSourceConfiguration): Promise<void>;
}

export interface MonitoredEmailService {
  listMonitoredEmails(): Promise<MonitoredEmailAddress[]>;
  addMonitoredEmail(email: string, providerHint: ProviderID): Promise<MonitoredEmailAddress>;
  setMonitoredEmailEnabled(id: string, isEnabled: boolean): Promise<void>;
  removeMonitoredEmail(id: string): Promise<void>;
}